// @adobe/ccweb-add-on-manifest/dist/manifest-field/AddOnManifestRequirement.js
/********************************************************************************
 * MIT License

 * © Copyright 2023 Adobe. All rights reserved.

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 ********************************************************************************/
import { ManifestVersion } from "../AddOnManifest.js";
import { AddOnManifestApp } from "./AddOnManifestApp.js";
/**
 * Defines the getter methods for Manifest Requirement field
 */
export class AddOnManifestRequirement {
    _manifestVersion;
    _requirements;
    _apps = [];
    constructor(_manifestVersion, _requirements) {
        this._manifestVersion = _manifestVersion;
        this._requirements = _requirements;
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                this._requirements = this._requirements;
                break;
            }
            default: {
                this._requirements = this._requirements;
            }
        }
    }
    // TODO: [WXP-1700] Return apps as AddOnManifestApp[] for all manifest versions
    get apps() {
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                return this._requirements.apps;
            }
            default: {
                if (!this._apps.length) {
                    this._requirements.apps.forEach(app => {
                        this._apps.push(new AddOnManifestApp(app));
                    });
                }
                return this._apps;
            }
        }
    }
    get experimentalApis() {
        return Boolean(this._requirements.experimentalApis);
    }
    get supportsTouch() {
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                return false;
            }
            default: {
                return Boolean(this._requirements.supportsTouch);
            }
        }
    }
    get renditionPreview() {
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                return false;
            }
            default: {
                return Boolean(this._requirements.renditionPreview);
            }
        }
    }
    get privilegedApis() {
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                return false;
            }
            default: {
                return Boolean(this._requirements.privilegedApis);
            }
        }
    }
    get _blessedPartnerAccess() {
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                return undefined;
            }
            default: {
                return this._requirements._blessedPartnerAccess;
            }
        }
    }
    get trustedPartnerApis() {
        switch (this._manifestVersion) {
            case ManifestVersion.V1: {
                return undefined;
            }
            default: {
                return this._requirements.trustedPartnerApis;
            }
        }
    }
}
